etuutt_bot/services/role.py

```diff
@@ -0,0 +1,119 @@
+from __future__ import annotations
+
+import itertools
+import operator
+from enum import Enum
+from functools import reduce
+from typing import TYPE_CHECKING, assert_never
+
+from discord import Permissions, Role
+
+if TYPE_CHECKING:
+    from etuutt_bot.bot import EtuUTTBot
+
+
+class MergeStrategy(Enum):
+    """Stratégie pour la fusion de plusieurs objets."""
+
+    Intersection = 1
+    """On garde l'intersection. ex: `[a, b, c] & [b, c, d] => [b, c,]`."""
+    Union = 2
+    """On garde l'union. ex: `[a, b, c] | [b, c, d] => [a, b, c, d]`."""
+    Clear = 3
+    """On ne garde rien. ex: `[a, b, c] | [b, c, d] => []`"""
+
+
+class RoleService:
+    """Service de gestion des rôles du serveur."""
+
+    def __init__(self, bot: EtuUTTBot):
+        self._bot = bot
+
+    def get_duplicates(self, *, case_sensitive: bool = True) -> list[list[Role]]:
+        """Cherche et retourne tous les rôles qui sont dupliqués.
+
+        Returns:
+            La liste des listes de duplication.
+            Chaque élément de la liste est une liste dont tous les
+            éléments sont des rôles dupliqués.
+        """
+
+        def sort_key(r):
+            return r.name if case_sensitive else r.name.lower()
+
+        roles = iter(sorted(self._bot.watched_guild.roles, key=sort_key))
+
+        res = []
+        for _, group in itertools.groupby(roles):
+            group = list(group)
+            if len(group) > 1:
+                res.append(group)
```

Comment:
```suggestion
        for _, group in itertools.groupby(roles, key=sort_key):
            values = list(group)
            if len(values) > 1:
                res.append(values)
```
Et je pense que ton intention était de group avec la même clé que pour sort (conseillé dans la doc de groupby). Et tu redéfinis group, je sais pas si c'est gênant mais c'est pas mieux de le nommer autrement ?